src/FakeItEasy.Analyzer/NonVirtualSetupAnalyzer.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace FakeItEasy.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class NonVirtualSetupAnalyzer:DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = CreateSupportedDiagnostics();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeCall, SyntaxKind.InvocationExpression);
+        }
+
+        private static void AnalyzeCall(SyntaxNodeAnalysisContext context)
+        {
+            var invocationExpression = (InvocationExpressionSyntax)context.Node;
+
+            if (IsInterface(context, invocationExpression))
+                return;
+
+            if (NotSetupCall(invocationExpression))
+                return;
+
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(context.Node);
+
+            if (!symbolInfo.Symbol.IsVirtual)
+            {
+                var location = invocationExpression.GetLocation();
+
+                Diagnostic diagnostic = Diagnostic.Create(DiagnosticDefinitions.NonVirtualSetup, location, symbolInfo.Symbol.Name);
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static bool NotSetupCall(InvocationExpressionSyntax invocationExpression)

```

Comment:
A nit, but if you're in the area again, our preference is to prefix Boolean properties and methods with "Is/Can/Has", as you've done with `IsInterface`. 

(And my personal preference is to lean toward affirmative phrases, such as `IsSetupCall`, but in this case, given the limited usage, I think your choice of the negative makes a better call site: `if (IsNotSetupCall` vs. `if (!IsSetupCall`. Alternative would be to invert the `if` so that you had `if ( IsSetupCall… ) { … }`, but that introduces nesting, so it's tough. I'm not asking for any action in this digression; it's just food for thought.)